### PR TITLE
[Compatibility] Add `Module#refinements`

### PR DESCRIPTION
**Pull request by itarato**
Source: https://github.com/oracle/truffleruby/issues/3039

Module#refinements has been added. [[Feature #12737](https://bugs.ruby-lang.org/issues/12737)]

### CODE REVIEW

File: test/mri/tests/ruby/test_refinement.rb [hunk 1]
```diff
@@ -1835,6 +1835,21 @@ def test_used_modules
     assert_equal [ref::RefB, ref::RefA], ref::Combined::USED_MODS
   end
 
+  def test_refinements
```
Comment (itarato) [inline test/mri/tests/ruby/test_refinement.rb:1838]: Copied from: https://bugs.ruby-lang.org/projects/ruby-master/repository/git/revisions/54198c7b97d3d353f7ac233e0360034b6e7b6cb6/entry/test/ruby/test_refinement.rb#L1751

Comment (eregon) [reply test/mri/tests/ruby/test_refinement.rb:1838]: Test comment for Slack notifications

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java [hunk 2]
```diff
@@ -2335,15 +2334,23 @@ protected RubyArray usedRefinements() {
...
         }
 
     }
 
+    @CoreMethod(names = "refinements")
+    public abstract static class RefinementsNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyArray refinements(RubyModule self) {
+            return createArray(self.fields.getRefinements().values().toArray());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleNodes.java:2349]: Needs a TruffleBoundary (on the specialization to keep it simple) for the `.values()` and `.toArray()`

Comment (itarato) [reply src/main/java/org/truffleruby/core/module/ModuleNodes.java:2350]: Added.

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
An MRI test is failing:
```
  1) Failure:
TestRefinement#test_refinements [/b/b/e/main/test/mri/tests/ruby/test_refinement.rb:1850]:
<[#<refinement:Integer@#<Module:0x115a3b8>>,
 #<refinement:String@#<Module:0x115a3b8>>]> expected but was
<[#<refinement:String@#<Module:0x115a3b8>>,
 #<refinement:Integer@#<Module:0x115a3b8>>]>.
```

**Comment by itarato** [comment]
> An MRI test is failing (we should try to run those on GitHub Actions):
> 
> ```
>   1) Failure:
> TestRefinement#test_refinements [/b/b/e/main/test/mri/tests/ruby/test_refinement.rb:1850]:
> <[#<refinement:Integer@#<Module:0x115a3b8>>,
>  #<refinement:String@#<Module:0x115a3b8>>]> expected but was
> <[#<refinement:String@#<Module:0x115a3b8>>,
>  #<refinement:Integer@#<Module:0x115a3b8>>]>.
> ```

@eregon To respect ordering I switched to linked hash map, which has I believe has a minor disadvantage on create (in time) as well as I believe using atomicity via `Collections.synchronizedMap` is slighly slower than `ConcurrentHashMap`. Would we consider this acceptable or should we look for alternatives?

**Comment by eregon** [comment]
There are two problems with that:
* `Collections.synchronizedMap()` does not allow parallel lookups. And so it means method lookup in different threads would cause contention if there are refinements active in the scope where the lookup happens.
* `Collections.synchronizedMap()` does not synchronize iterables like `for (var entry : map.entrySet())`, which is then incorrect synchronization and causes all sorts of problems.

I think the test should allow both, ordering there seems artificial and unnecessary, it will not affect method lookup semantics, as anyway it's a single refinement module per class/module being refined inside a namespace module.
Could you either tag the MRI test or not add it in this PR? There is already a spec doing the appropriate check.

**Review by eregon — APPROVED** [review]

**Pull request merged**